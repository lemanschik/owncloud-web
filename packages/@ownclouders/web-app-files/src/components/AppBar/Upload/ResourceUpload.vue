<template>
  <div>
    <oc-button :class="btnClass" justify-content="left" appearance="raw" @click="triggerUpload">
      <oc-resource-icon :resource="resource" size="medium" />
      <span :id="uploadLabelId">{{ buttonLabel }}</span>
    </oc-button>
    <input
      :id="inputId"
      ref="input"
      v-bind="inputAttrs"
      class="upload-input-target"
      type="file"
      :aria-labelledby="uploadLabelId"
      :name="isFolder ? 'file' : 'folder'"
      tabindex="-1"
    />
  </div>
</template>

<script lang="ts">
import { computed, defineComponent } from 'vue'
import { Resource } from 'web-client/src'
export default defineComponent({
  props: {
    btnLabel: {
      type: String,
      required: false,
      default: ''
    },
    btnClass: {
      type: String,
      required: false,
      default: ''
    },
    isFolder: {
      type: Boolean,
      required: false,
      default: false
    }
  },
  setup(props) {
    const resource = computed(() => {
      return { extension: '', isFolder: props.isFolder } as Resource
    })
    return {
      resource
    }
  },
  computed: {
    inputId() {
      if (this.isFolder) {
        return 'files-folder-upload-input'
      }
      return 'files-file-upload-input'
    },
    uploadLabelId() {
      if (this.isFolder) {
        return 'files-folder-upload-button'
      }
      return 'files-file-upload-button'
    },
    buttonLabel() {
      if (this.btnLabel) {
        return this.btnLabel
      }
      if (this.isFolder) {
        return this.$gettext('Folder')
      }
      return this.$gettext('Files')
    },
    inputAttrs() {
      if (this.isFolder) {
        return {
          webkitdirectory: true,
          mozdirectory: true,
          allowdirs: true
        }
      }
      return { multiple: true }
    }
  },
  mounted() {
    this.$uppyService.registerUploadInput(this.$refs.input as HTMLInputElement)
  },
  beforeUnmount() {
    this.$uppyService.removeUploadInput(this.$refs.input as HTMLInputElement)
  },
  methods: {
    triggerUpload() {
      ;(this.$refs.input as HTMLInputElement).click()
    }
  }
})
</script>

<style scoped>
.upload-input-target {
  position: absolute;
  left: -99999px;
}
</style>
