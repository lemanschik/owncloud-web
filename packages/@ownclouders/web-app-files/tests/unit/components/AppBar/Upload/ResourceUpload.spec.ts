import ResourceUpload from '../../../../../src/components/AppBar/Upload/ResourceUpload.vue'
import {
  createStore,
  defaultComponentMocks,
  defaultPlugins,
  defaultStoreMockOptions,
  defaultStubs,
  mount
} from 'web-test-helpers'

const Translate = jest.fn()

describe('Resource Upload Component', () => {
  describe('file upload', () => {
    it('should render component', () => {
      const { wrapper } = getWrapper()
      expect(wrapper.exists()).toBeTruthy()
      expect(wrapper.html()).toMatchSnapshot()
    })
  })

  describe('folder upload', () => {
    it('should render component', () => {
      const { wrapper } = getWrapper({ isFolder: true })
      expect(wrapper.exists()).toBeTruthy()
      expect(wrapper.html()).toMatchSnapshot()
    })
  })

  describe('when upload file button is clicked', () => {
    it('should call "triggerUpload"', async () => {
      const { wrapper } = getWrapper()

      const spyTriggerUpload = jest.spyOn(wrapper.vm, 'triggerUpload')
      const uploadButton = wrapper.find('button')
      const fileUploadInput = wrapper.find('#files-file-upload-input')

      ;(fileUploadInput.element as HTMLElement).click = jest.fn()
      await wrapper.vm.$forceUpdate()

      await uploadButton.trigger('click')

      expect(spyTriggerUpload).toHaveBeenCalledTimes(1)
      expect((fileUploadInput.element as HTMLElement).click).toHaveBeenCalledTimes(1)
    })
  })
})

function getWrapper(props = {}) {
  const storeOptions = { ...defaultStoreMockOptions }
  const store = createStore(storeOptions)
  const mocks = defaultComponentMocks()
  return {
    storeOptions,
    mocks,
    wrapper: mount(ResourceUpload, {
      props,
      global: {
        mocks,
        stubs: defaultStubs,
        plugins: [...defaultPlugins(), store],
        directives: { Translate }
      }
    })
  }
}
