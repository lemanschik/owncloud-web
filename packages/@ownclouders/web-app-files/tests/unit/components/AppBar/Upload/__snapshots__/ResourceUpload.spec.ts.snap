// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Resource Upload Component file upload should render component 1`] = `
<div>
  <button class="oc-button oc-rounded oc-button-m oc-button-justify-content-left oc-button-gap-m oc-button-passive oc-button-passive-raw" type="button">
    <!-- @slot Content of the button -->
    <span class="oc-icon oc-icon-m oc-icon-passive oc-resource-icon oc-resource-icon-file">
      <!---->
    </span>
    <span id="files-file-upload-button">Files</span>
  </button>
  <input aria-labelledby="files-file-upload-button" class="upload-input-target" id="files-file-upload-input" multiple="" name="folder" tabindex="-1" type="file">
</div>
`;

exports[`Resource Upload Component folder upload should render component 1`] = `
<div>
  <button class="oc-button oc-rounded oc-button-m oc-button-justify-content-left oc-button-gap-m oc-button-passive oc-button-passive-raw" type="button">
    <!-- @slot Content of the button -->
    <span class="oc-icon oc-icon-m oc-icon-passive oc-resource-icon oc-resource-icon-folder">
      <!---->
    </span>
    <span id="files-folder-upload-button">Folder</span>
  </button>
  <input allowdirs="true" aria-labelledby="files-folder-upload-button" class="upload-input-target" id="files-folder-upload-input" mozdirectory="true" name="file" tabindex="-1" type="file" webkitdirectory="true">
</div>
`;
